drivers/cartesi/cmio.c

```diff
@@ -0,0 +1,231 @@
+	return container_of(dev, struct cmio_device, mdev);
+}
+
+static long cmio_ioctl_setup(struct cmio_device *me, unsigned long arg)
+{
+	if (copy_to_user((void __user *)arg, &me->bufs, sizeof me->bufs))
+		return -EFAULT;
+
+	return 0;
+}
+
+static long cmio_ioctl_yield(struct cmio_device *me, unsigned long arg)
+{
+	__u64 req = 0,
+	      rep = 0;
+
+	if (copy_from_user(&req, (void __user *)arg, sizeof req))
+		return -EFAULT;
+
+	rep = sbi_ecall(SBI_YIELD, 0, req, 0, 0, 0, 0, 0).value;
+
+	if (copy_to_user((void __user *)arg, &rep, sizeof rep))
+		return -EFAULT;
+
+	return 0;
+}
+
+/*
+ * We enforce only one user at a time here with the open/release.
+ */
+static int cmio_open(struct inode *inode, struct file *file)
+{
+	struct cmio_device *cmio = to_cmio_device(file);
+	if (!cmio)
+		return -EBADF;
+
+	if (!atomic_dec_and_test(&cmio->single_user_lock)) {
+		atomic_inc(&cmio->single_user_lock);
+		return -EBUSY;
+	}
+	return 0;
+}
+
+static int cmio_release(struct inode *inode, struct file *file)
+{
+	struct cmio_device *cmio = to_cmio_device(file);
+	if (!cmio)
+		return -EBADF;
+
+	atomic_inc(&cmio->single_user_lock);
+	return 0;
+}
+
+static long cmio_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
+{
+	struct cmio_device *cmio = to_cmio_device(file);
+	if (!cmio)
+		return -EBADF;
+
+	switch (cmd) {
+	case IOCTL_CMIO_SETUP:
+		return cmio_ioctl_setup(cmio, arg);
+	case IOCTL_CMIO_YIELD:
+		return cmio_ioctl_yield(cmio, arg);
+	}
+	return -ENOIOCTLCMD;
+}
+
+static int cmio_mmap(struct file *file, struct vm_area_struct *vma)
+{
+	__u64 vma_size;
+	struct cmio_device *cmio = to_cmio_device(file);
+	if (!cmio)
+		return -EBADF;
+
+	vma_size = vma->vm_end - vma->vm_start;
+	if (!((vma->vm_start == cmio->bufs.tx.data && vma_size == cmio->bufs.tx.length)
+	||    (vma->vm_start == cmio->bufs.rx.data && vma_size == cmio->bufs.rx.length)))
+		return -EINVAL;
+
+	return remap_pfn_range(vma,
+	                       vma->vm_start,
+	                       vma->vm_start >> PAGE_SHIFT,
+	                       vma_size,
+	                       vma->vm_page_prot);
+}
+
+static const struct file_operations cmio_fileops = {
+	.open           = cmio_open,
+	.release        = cmio_release,
+	.unlocked_ioctl = cmio_ioctl,
+	.mmap           = cmio_mmap,
+	.owner          = THIS_MODULE,
+};
+
+static int setup_buffer(struct device_node *parent
+                       ,const char *name
+                       ,struct cmio_buffer *buf)
+{
+	u64 xs[2];
+	int rc = -EIO;
+	struct device_node *node = NULL;
+
+	if (!(node = of_find_node_by_name(parent, name)))
+		goto leave;
+	if (of_property_read_u64_array(node, "reg", xs, 2))
+		goto leave;
+	if (!(buf->data = xs[0]))
+		goto leave;
+	if (!(buf->length = xs[1]))
+		goto leave;
+	rc = 0;
+leave:
+	of_node_put(node);
+	return rc;
+}
+
+static int setup_io(struct cmio_device *cmio)
+{
+	int rc = -EIO;
+	struct device_node *cmio_node = NULL;
+
+	if (!(cmio_node = of_find_node_by_path("/rollup"))
```

Comment:
Are we still using the name "rollup"?

Reply:
That is the node name on the `FDT` (emulator: dtb.cpp). I can change it to `cmio` as well.

Reply:
I think it makes sense to do that everywhere.